### PR TITLE
Remove the duplicate key/value when ingesting log from opentelemetry-dotnet

### LINKED ISSUE #3868: [BUG] Duplicate key exception when integrate with open-telemetry exporter
**Describe the bug**
Data prepper throws java.lang.IllegalStateException: Duplicate key xxx when ingest through log pipeline. 
See the error below, due to open telemetry exporter exporting a duplicate **ConnectionId** key. Not sure this is expected behavior or not. Please help considering to fix this issue. 
```js
2023-12-29T07:28:02,516 [pool-13-thread-1] ERROR org.opensearch.dataprepper.plugins.source.otellogs.OTelLogsGrpcService - Failed to parse the request resource_logs {
  resource {
    ...truncate...
  scope_logs {
    scope {
      name: "Microsoft.Extensions.Hosting.Internal.Host"
    }
    log_records {
      time_unix_nano: 1703834880595497100
      severity_number: SEVERITY_NUMBER_DEBUG
      severity_text: "Debug"
      body {
        string_value: "Connection id \"{ConnectionId}\" sending FIN because: \"{Reason}\""
      }
      attributes {
        key: "ConnectionId"
        value {
          string_value: "0HN087UNTCNA9"
        }
      }
      attributes {
        key: "Reason"
        value {
          string_value: "The Socket transport\'s send loop completed gracefully."
        }
      }
      attributes {
        key: "ConnectionId"
        value {
          string_value: "0HN087UNTCNA9"
        }
      }
      observed_time_unix_nano: 1703834880595497100
    }
  }
}
java.lang.IllegalStateException: Duplicate key log.attributes.ConnectionId (attempted merging values 0HN087UNTCNA9 and 0HN087UNTCNA9)
	at java.base/java.util.stream.Collectors.duplicateKeyException(Collectors.java:135) ~[?:?]
	at java.base/java.util.stream.Collectors.lambda$uniqKeysMapAccumulator$1(Collectors.java:182) ~[?:?]
	at java.base/java.util.stream.ReduceOps$3ReducingSink.accept(ReduceOps.java:169) ~[?:?]
	at java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1625) ~[?:?]
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509) ~[?:?]
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499) ~[?:?]
	at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:921) ~[?:?]
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234) ~[?:?]
	at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:682) ~[?:?]
	at org.opensearch.dataprepper.plugins.otel.codec.OTelProtoCodec.unpackKeyValueListLog(OTelProtoCodec.java:1094) ~[otel-proto-common-2.6.1.jar:?]
	at org.opensearch.dataprepper.plugins.otel.codec.OTelProtoCodec$OTelProtoDecoder.lambda$processLogsList$7(OTelProtoCodec.java:376) ~[otel-proto-common-2.6.1.jar:?]
```
**To Reproduce**
1. I have pushed the sample repository, please visit [here](https://github.com/hoangthanh28/data-prepper) 
2. Run `docker-compose -f docker-compose.yml up -d`
3. `curl --location 'http://localhost:5000'`
4. Wait a couple of seconds, then check data-prepper log. `docker logs -f data-prepper`

**Expected behavior**
Remove the duplicate key and ingest the log without issue.
Not sure whether this is a right place to fix or not ( forgive me if I'm wrong, I'm a .net developer - FYI)
[Code](https://github.com/opensearch-project/data-prepper/blob/f19de03d5418925935e019837fc4824fb250820c/data-prepper-plugins/otel-proto-common/src/main/java/org/opensearch/dataprepper/plugins/otel/codec/OTelProtoCodec.java#L376)
```java
 protected List<OpenTelemetryLog> processLogsList(final List<LogRecord> logsList,
                                                                                         final String serviceName,
                                                                                         final Map<String, Object> ils,
                                                                                         final Map<String, Object> resourceAttributes,
                                                                                         final String schemaUrl) {
            return logsList.stream()
                    .map(log -> JacksonOtelLog.builder()
                            .withTime(OTelProtoCodec.convertUnixNanosToISO8601(log.getTimeUnixNano()))
                            .withObservedTime(OTelProtoCodec.convertUnixNanosToISO8601(log.getObservedTimeUnixNano()))
                            .withServiceName(serviceName)
                            .withAttributes(OTelProtoCodec.mergeAllAttributes(
                                    Arrays.asList(
                                            // would be possible to de-duplicate log.getAttributesList() first ???, so that will solve the problem.
                                            OTelProtoCodec.unpackKeyValueListLog(log.getAttributesList()),
                                            resourceAttributes,
                                            ils
                                    )
                            ))
                            .withSchemaUrl(schemaUrl)
                            .withFlags(log.getFlags())
                            .withTraceId(OTelProtoCodec.convertByteStringToString(log.getTraceId()))
                            .withSpanId(OTelProtoCodec.convertByteStringToString(log.getSpanId()))
                            .withSeverityNumber(log.getSeverityNumberValue())
                            .withSeverityText(log.getSeverityText())
                            .withDroppedAttributesCount(log.getDroppedAttributesCount())
                            .withBody(OTelProtoCodec.convertAnyValue(log.getBody()))
                            .build())
                    .collect(Collectors.toList());
        }
```
**Screenshots**

**Environment (please complete the following information):**
 - OS: Debian
 - Version: 12
 
**Additional context**
N/A

### PR DESCRIPTION
**Pull request by hoangthanh28**
### Description
This PR is intended to resolve the duplicate key exception while ingesting the otel-dotnet log.
 
### Issues Resolved
Resolves #3868
 
### Check List
- [ ] New functionality includes testing.
- [ ] New functionality has a documentation issue. Please link to it in this PR.
  - [ ] New functionality has javadoc added
- [x] Commits are signed with a real name per the DCO

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/data-prepper/blob/main/CONTRIBUTING.md).

### CODE REVIEW

**Review by KarstenSchnitter — COMMENTED** [review]
Change looks good to me. I wonder, whether we want tests for that. But this is just Java Streams API. @hoangthanh28, check your commit messages for correct DCO: "signed-off-by: ..."

**Comment by dlvenable** [comment]
@hoangthanh28 , Thank you for this contribution! I agree with @KarstenSchnitter that we should have a unit test. You can write a test which sends in multiple values and make sure you get the last one.

File: data-prepper-plugins/otel-proto-common/src/test/java/org/opensearch/dataprepper/plugins/otel/codec/OTelProtoCodecTest.java [hunk 1]
```diff
@@ -1318,4 +1318,61 @@ public void testBoundsKeyNotEqualsSign() {
...
+        final String keyValue = "duplicate_value";
+        final KeyValue duplicateAttribute1 = KeyValue.newBuilder().setKey(keyName).setValue(AnyValue.newBuilder()
+                .setStringValue(keyValue).build()).build();
+        final KeyValue duplicateAttribute2 = KeyValue.newBuilder().setKey(keyName).setValue(AnyValue.newBuilder()
+                .setStringValue(keyValue).build()).build();
+        final NumberDataPoint actual = NumberDataPoint.newBuilder()
+                .addAllAttributes(Arrays.asList(duplicateAttribute1, duplicateAttribute2)).build();
+        
+        Map<String, Object> map = OTelProtoCodec.convertKeysOfDataPointAttributes(actual);
+        assertThat(map.size(), is(equalTo(1)));
```
Comment (KarstenSchnitter) [inline data-prepper-plugins/otel-proto-common/src/test/java/org/opensearch/dataprepper/plugins/otel/codec/OTelProtoCodecTest.java:1333]: I understand, that it is enough to test for non-duplication. However, this is unspecified behaviour, when the key values are different. It is not clear, which value will be chosen. Even if this is undocumented behaviour, I think the unit test should force a specific selection. This is how I understand @dlvenable comment: https://github.com/opensearch-project/data-prepper/pull/4394#issuecomment-2039952138

**Review by KarstenSchnitter — COMMENTED** [review]
Thanks for providing the change and the unit tests. I am not sure, whether the test should enforce the decision for the first occurring key-value pair over potential later duplicates.

**Comment by hoangthanh28** [comment]
Merge from original repo

**Pull request closed without merging**